Generic unit for surrogate models or custom models

jghouse88 opened the pull request:
Thanks to @carldlaird , @eslickj, @andrewlee94 and @jsiirola for helpful feedback. I think this is ready for review now. I renamed the unit to `SkeletonUnitModel` based on comments. If anyone has a better name, I am open to suggestions. 

## Summary/Motivation:


- There is a need for having an IDAES unit model object that can be used to add a surrogate for an entire unit with certain inputs/outputs or to have a simple custom block that allows the user to add any custom implementation for a model as they see fit. Currently, we do not have an off-the-shelf solution from the unit_models library that lets us do this. We need something like this in several projects now. 
- While a Pyomo Block can be used for this, linking with other IDAES models is limited with the absence of inlet and outlet ports. This lets us have an API that can work with other parts of IDAES.
- UnitModel could have been used but there is a lot of infrastructure built in there that is not required for this and hence this inherits from ProcessBlock. Also, the initialization method simply calls the controlvolume intialize method which will not be available for surrogate or custom models. 


## Changes proposed in this PR:
- Ability to add multiple inlets and outlets and populate with whatever vars user wants. Linking with other unit models in IDAES will probably need a translator block to link variables in ports or won't be necessary if the surrogate uses same state vars for inputs and outputs. 
- Default initialize method and an ability for user to use a callback if custom initialization is required
- See `test_skeleton_unit_model.py` for an example
- Added documentation

## What this could develop into?
- Better integration with our surrogate modeling tools such as ALAMO/PySMO. Ability to directly pass an output json or lst file as a  config argument to build an entire surrogate rather than user manually adding set of vars and constraints. Note this is different from simply adding an expression for a property calculation. 
- Integrate with ideas proposed by @eslickj to look at constrained regression with mass/energy balance. 
- IDAES 2.0 is looking at integration with ML tools in Python and we can create the API to integrate with such models.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

eslickj commented on idaes/generic_models/unit_models/__init__.py:
```diff
@@ -16,3 +16,4 @@
 from .statejunction import StateJunction
 from .translator import Translator
 from .valve import ValveFunctionType, Valve, ValveData
+from .surrogate_model import SurrogateModel
```
May as well import SurrogateModelData here too if you may want to subclass the surrogate block.  If not disregard.

eslickj commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+
+        Keyword Args:
+            name : name to use for Port object.
+            dict : dictionary containing variables to be added to the port
```
Is dict member_list?  If dict is member_list, them maybe member_dict would be more clear.

eslickj commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+            p.add(Reference(local_name), k)
+
+    def initialize(self, custom_initialize=None, outlvl=idaeslog.NOTSET,
+                   solver='ipopt', optarg={'tol': 1e-6}):
```
Probably go with optarg=None as default.  @andrewlee94's get_solver() method will get a solver with the default tolerance of 1e-6.  I think it's the new standard in initialization.

andrewlee94 replied:
On that line, the default value for `solver` is now `None` as well, as `get_solver()` will return the configured default solver in this case.

eslickj commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+        solve_log = idaeslog.getSolveLogger(self.name, outlvl, tag="unit")
+
+        opt = SolverFactory(solver)
+        opt.options = optarg
```
I think you can replace these two lines with

```python
if optarg is None:
    optarg = {}
opt = get_solver(solver, oparg)
```

get_solver can be imported from idaes.core.util.misc

andrewlee94 replied:
`get_solver` now accepts `optarg=None` as an argument, so this can be done a a single line:

```
opt = get_solver(solver, optarg)
```

eslickj commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+        opt.options = optarg
+
+        if custom_initialize is not None:
+            custom_initialize
```
You'll want to call the callback: ```custom_initialize()```

Daison2102 commented:
Started testing this class to implement a custom DAC model in IDAES and it was simple to use and very helpful.

andrewlee94 commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+        setattr(self, name, p)
+
+        # Create References for port members
+        for k in member_list.keys():
```
This will depend a bit on how the surrogate itself is built, but do we need to make `References` here?

This was necessary in the core models to make it look lie the state variables were directly indexed by time (instead of being on a `Block` indexed by time). Here I would have guessed that any variable would be directly indexed by time (as surrogates probably lack the block-hierarchical structure of other models). If that is the case, then we should skip the `References`.

andrewlee94 replied:
An important thing to remember here is that `Port` members need to be indexed by time in order to link up with other IDAES models.

That means that either the internal surrogate model variables need to be indexed by time, or we need to add a time index to them in the `Ports` (via `References`). We need to decide which approach we are going to use and clearly document it.

jghouse88 replied:
@andrewlee94 Thanks for this comment. I am leaning towards model variables indexed by time and leaving it to user. See test for an example. I have also documented this.

andrewlee94 replied:
@jghouse88 I think indexing the variables (and surrogate constraints) by time is the best approach as well, and the draft surrogate API supports this. This makes it a lot easier at the `SkeletonModel` level as you can just add variables directly to `Ports` as required.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+from idaes.core import FlowsheetBlock
+from idaes.generic_models.unit_models import SurrogateModel
+from idaes.core.util.model_statistics import degrees_of_freedom
+from idaes.core.util.testing import get_default_solver
```
Should switch to `get_solver` now.

`get_default_solver` just points to that now with a deprecation warning.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+
+# -----------------------------------------------------------------------------
+# Get default solver for testing
+solver = get_default_solver()
```
As above.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+solver = get_default_solver()
+# -----------------------------------------------------------------------------
+
+m = ConcreteModel()
```
It would be a good idea to wrap this in a fixture so that it isn't run on import.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+    m.fs.surrogate.x_1.fix(1)
+    m.fs.surrogate.x_2.fix(2)
+
+    opt = SolverFactory("ipopt")
```
Why not use the `solver` object created earlier?

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+
+    opt = SolverFactory("ipopt")
+
+    opt.solve(m)
```
Might it be worth recording the solve status of both these solves and  attaching it to the model so that you can asset solver status during the tests (to make sure things worked as expected).

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+    opt.solve(m)
+
+
+def test_ports():
```
Need pytest marks.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_surrogate_model.py:
```diff
@@ -0,0 +1,118 @@
+
+    assert value(m.fs.surrogate.y_1) == pytest.approx(0.5, abs=1e-3)
+    assert value(m.fs.surrogate.y_2) == pytest.approx(0.25, abs=1e-3)
+
```
Should test a case for the default initialize with too few degrees of freedom and check that the expected error is raised.

andrewlee94 reviewed (commented):
Some comments and suggestions.

A big part of this is going to have to be documenting what a surrogate model is expected to look like inside (i.e. what the user needs to add), such as how variables should be indexed (especially regarding time indices and `Port` members).

jghouse88 commented:
@eslickj and @andrewlee94 Thank you for your comments. I will address them and wait for any other feedback from others. 

@Daison2102 Thanks for testing this with the DAC model. @minhwvu you can use the code from my branch until it is made available in IDAES.

minhwvu commented:
> @eslickj and @andrewlee94 Thank you for your comments. I will address them and wait for any other feedback from others.
> 
> @Daison2102 Thanks for testing this with the DAC model. @minhwvu you can use the code from my branch until it is made available in IDAES.

Hi @jghouse88 : I run a test and it works for my model. Thanks.

jghouse88 commented on idaes/generic_models/unit_models/surrogate_model.py:
```diff
@@ -0,0 +1,135 @@
+
+    # Create Class ConfigBlock
+    CONFIG = ProcessBlockData.CONFIG()
+    CONFIG.declare("dynamic", ConfigValue(
```
See next comment on time.

andrewlee94 commented on idaes/generic_models/unit_models/skeleton_model.py:
```diff
@@ -0,0 +1,154 @@
+        default=False,
+        domain=In([False]),
+        description="Dynamic model flag",
+        doc="""Model does not handle dynamics but should be set by user."""))
```
Not quite sure what this doc string means. The first part says it doesn't, but the second part says the user should set it?

Also, a related note, the absence of `has_holdup` means that `ProcessBlockData._setup_dynamics` will fail (as it expects `has_holdup` to be present). I do not think this is a big problem here (at least not yet), but we should be a ware of this at least. We might even want to consider overloading `_setup_dynamics` with a `NotSupported` error.

jghouse88 replied:
Updated the doc string to better explain what I meant. 

Regarding the second comment, actually it fails without the dynamic flag.

andrewlee94 commented on idaes/generic_models/unit_models/skeleton_model.py:
```diff
@@ -0,0 +1,154 @@
+        for k in member_dict.keys():
+            p.add(member_dict[k], name=k)
+
+    def _custom_initialize_method(self):
```
This seems to be a round-about way of doing this, and there is not an obvious API for how the user would setthe custom initialize method.

My first instinct is to just put a reference to the custom initialization method in the config block, and call it directly from there. I.e.

```
def initialize(self, *args, **kwargs):
    if self.config.custom_initializer is None:
        self._default_initialize(*args, **kwargs)
    else:
        self.config.custom_initializer(*args, **kwargs)
```

There is also a question of defining the standard arguments a custom initialization methods should expect.

jsiirola replied:
I think I would go one farther and just call the config entry `initializer`, and have it hold a callable.  The default value would be `SkeletonUnitModelData._default_initializer`.  I think this would work easier if the function signature was `(unit, *args, **kwargs)`, then `_default_initializer` could be a `@staticmethod` (and declared above the `CONFIG` declaration).  That would help with things like cloning / pickling, as the callable is no longer a bound method.

jghouse88 replied:
Thanks for the suggestions @andrewlee94  and @jsiirola. I have updated as recommended. Thanks for @jsiirola for help with the staticmethod. I think it looks nice and clean now.

andrewlee94 commented on idaes/generic_models/unit_models/skeleton_model.py:
```diff
@@ -0,0 +1,154 @@
+
+        if self.config.custom_initializer:
+            self._custom_initialize_method()
+        elif degrees_of_freedom(self) == 0:
```
Should this be a global check? I.e. shouldn't this also apply even if there is a custom initialisation method?

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_skeleton_unit_model.py:
```diff
@@ -0,0 +1,305 @@
+                       "temperature": m.fs.skeleton.temperature_out,
+                       "pressure": m.fs.skeleton.pressure_out}
+
+        m.fs.skeleton.add_ports(name="inlet", member_dict=inlet_dict)
```
Just from the view point of unit testing methods, I would suggest only calling these in the `test_ports` method. That way the method is being called as part of the test for its effects.

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_skeleton_unit_model.py:
```diff
@@ -0,0 +1,305 @@
+    def test_default_initialize(self, skeleton_default):
+        # Check default initialize method raises ConfigurationError
+
+        with pytest.raises(ConfigurationError):
```
It is good practice to check the error message here to make sure you are not getting another unexpected error.

```
with pytest.raises(ConfigurationError, match="expected string"):
```

andrewlee94 commented on idaes/generic_models/unit_models/tests/test_skeleton_unit_model.py:
```diff
@@ -0,0 +1,305 @@
+        with pytest.raises(ConfigurationError):
+            skeleton_default.fs.skeleton.initialize()
+
+        skeleton_default.fs.skeleton.inlet.flow_mol_comp[0, "c1"].fix(2)
```
I would also break this into two tests, just so failure in the first part does not obscure any failures in the second part.

andrewlee94 reviewed (commented):
I didn't read the docs thoroughly at this point, but the rest looks pretty good.

My biggest comment is that the way you set up the custom initialization method seems rather cumbersome and doesn't offer an obvious way for the user to provide the custom method. I think it would be much more straight-forward to provide the custom method as an input argument in the config block (rather than the current `bool`) and call it directly from the config block. This way there is an obvious API for user on how they should provide the method.

jghouse88 commented:
This is ready for review.

andrewlee94 commented on docs/technical_specs/model_libraries/generic/unit_models/skeleton_unit.rst:
```diff
@@ -0,0 +1,67 @@
+
+The IDAES Skeleton Unit model represents a bare bones unit (hence, the name skeleton) that lets the user define a unit model with a custom set
+of variables and constraints that best suits their needs but that can be connected to other models from the unit model library. This unit can be used
+to either define a surrogate for a complex unit model or to define a first-principles based model that does not use IDAES
```
Minor point, but there is nothing stopping a user from adding ControlVolumes here either.

jghouse88 replied:
Yes, they can but I do not see why..

andrewlee94 replied:
My point being we should not give the impression of restrictions that do not actually exist. I would make it more vague, such as "first principles models that do not adhere to the typical IDAES modeling hierarchy".

andrewlee94 commented on docs/technical_specs/model_libraries/generic/unit_models/skeleton_unit.rst:
```diff
@@ -0,0 +1,67 @@
+Degrees of Freedom
+------------------
+
+There are no variables or constraints as this is a bare minimum skeleton that will be populated with variables and constraints
```
The following three sections aren't really applicable in this case (as everything depends on the user). It would probably be better to give an overview of how to use the skeleton model, and what the user needs to keep in mind.

Some examples:

* Port members need to be indexed by time. That leads into how to access the time domain (I will add a PR for making `time` a property of the flowsheet).
* How to use the methods to add ports and how to define the members of the port.
* A reminder that Ports must be consistent, thus users need to match their Port definitions to the flowsheet they intend to use it in.

andrewlee94 replied:
Two other things that would be worth mentioning:

* unit consistency (if the user wishes to assert unit consistency across the model, then the Skeleton will need to be unit consistent)
* Brief mention of how this could interact with the scaling tools (for now, just say the user would need to do this manually).

jghouse88 replied:
I plan to create an example in examples-pse

andrewlee94 commented on docs/technical_specs/model_libraries/generic/unit_models/skeleton_unit.rst:
```diff
@@ -0,0 +1,67 @@
+
+.. code-block:: Python
+
+   self.config.initializer(self, opt=opt, init_log=init_log, solve_log=solve_log)
```
It might be worth noting the expectation of a static method and the fact that self here will be the skeleton model object. Given the range of users we expect, it might even be worth showing some pseudo-code of how to initialize a variable.

jsiirola replied:
There is no need for a user's custom function to be a static method - a "global" (unbound) function works as well.  I agree that we should clarify semantics in the documentation prototype and not use `self` here.

jghouse88 replied:
Before the code line, I do mention that "In the ``initialize`` method, after checking that degrees of freedom is zero, the following is triggered:". 
@jsiirola I am not sure I follow why self is not appropriate here.

jsiirola replied:
@jghouse88: good point; you are right.

andrewlee94 commented on idaes/generic_models/unit_models/skeleton_model.py:
```diff
@@ -0,0 +1,156 @@
+
+
+    def initialize(self, outlvl=idaeslog.NOTSET,
+                   solver=None, optarg=None):
```
Should we add an optional argument for providing initial guesses? This would probably be passed directly to the initializer method, so users could do what they want with it.

jghouse88 replied:
Interesting suggestion, however, how do you think we should handle that argument with the _default_initialize() method? Traditionally, we have assumed the initial guess to be applicable for the inlet port members. In the custom model case, the inlet port could be a subset of the input vars required for the model. I can always set `state_args` to `None` and not use it in the default_initialize and this could let the user use that argument in their custom method as they see fit. Let me know your thoughts.

andrewlee94 replied:
The simplest approach would to be accept it but then ignore it. Given that we don't know what a user will put inside a SkeletonModel, this is probably the best approach.

An alternative would be to assume these are initial values for variables in the outlet and try to assign to the values of these, but I am not sure that adds much (especially seeing as the main use for these models is for surrogates which are generally simple explicit functions).

andrewlee94 replied:
To add a little more - the main goal here is to give users some flexibility to pass extra information to their custom initializers. Obviously, this is not applicable to the default initializer, so it should not expect any additional information, but needs to have a placeholder for the addition information which it just ignores.

jghouse88 replied:
Alright.

andrewlee94 reviewed (dismissed):
Looks really good.

My comments are all minor things on potential (but not necessary) improvements.

jsiirola commented on docs/technical_specs/model_libraries/generic/unit_models/skeleton_unit.rst:
```diff
@@ -0,0 +1,67 @@
+
+.. code-block:: Python
+
+   def custom_func(unit, **kwargs):
```
Should we be recommending `**kwargs`, or should we write out the full prototype that will be used to call this?  That is,
```python
def custom_func(unit,  opt, init_log, solve_log):
    pass
```

jghouse88 replied:
Now that you mention it, I should probably show both..

jghouse88 commented:
@andrewlee94 and @jsiirola can you please take a look? I addressed the comments on the docs and also added `initial_guess` to the initialize method that can potentially be passed to the callback if the user desires.
